hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -1130,12 +1137,14 @@ int SaraNcpClient::muxChannelStateCb(uint8_t channel, decltype(muxer_)::ChannelS
                 // Muxer stopped
                 self->disable();
                 self->connState_ = NcpConnectionState::DISCONNECTED;
```

Comment:
I think this is actually the main problem. We've changed the state manually, without generating the event and also preventing the NCP client to do so properly. Calling just `disable()` should be enough to trigger a call to `enable()` in the next loop invocation (in the `PppNcpNetif`), which in turn will call `off()`, which will call `ncpState()`, which will call `connectionState()`.

We may potentially do the same thing similarly in https://github.com/particle-iot/device-os/pull/2049/files#diff-b3f74fd1e22ae53aae264584d668ea1cL363

Reply:
What about the other UBLOX_NCP_PPP_CHANNEL case in the muxer callback:
```
if (self->connState_ != NcpConnectionState::DISCONNECTED) {
    self->connState_ = NcpConnectionState::CONNECTING;
}
```
How would you propose we change/update this?

Initially I did simply remove `self->connState_ = NcpConnectionState::DISCONNECTED;` from the callback, and it seemed to resolve the issue... but when I checked what state was being seen by `connectionState()` on the next call that produced a PPP state change event, it was `NcpConnectionState::CONNECTING`, not `DISCONNECTED`.  This didn't seem like the right way to resolve the issue.  We want the shut down of the muxer to trigger a disable/shut down of the NCP and also the PPP client.  Seeing a `CONNECTING` state didn't seem like the right way to do that.

Also, the ESP32 code calls the `connectionState()` function directly from the muxer callback.  That would be a nice way to fix it as well, but it didn't seem like it was thread safe from the comments in the SARA muxer callback.

Reply:
> Also, the ESP32 code calls the connectionState() function directly from the muxer callback. That would be a nice way to fix it as well, but it didn't seem like it was thread safe from the comments in the SARA muxer callback.

We can actually do that (why don't we do that)? It's safe to call with `NcpConnectionState::DISCONNECTED`, because that won't call into the muxer. Let's do that.

> How would you propose we change/update this?

```c++
connectionState(NcpConnectionState::DISCONNECTED);
```

>  but when I checked what state was being seen by connectionState() on the next call that produced a PPP state change event, it was NcpConnectionState::CONNECTING, not DISCONNECTED

I'm not really sure where `CONNECTING` would be coming from, as that state is only generated by processEvents/checkRegistrationState, which would be bypassed in disabled state or `registerNet()`. And anyway, that state will trigger the teardown of the PPP session.

I'd also vote for calling just `disable()` in `dataChannelWrite()` on error conditions.